server/routes.js

```diff
@@ -19,3 +20,51 @@ export function topVolume (client) {
     res.json({ messages })
   }
 }
+
+export function topVolumeMatrix (client) {
+  return async (req, res) => {
+    let { from, to } = req.query
+    const maxHeight = await queries.maxHeight(client)
+    if (!to || !from) {
+      from = from || maxHeight
+      to = to || maxHeight + 1
+    }
+    const messages = await queries.topVolume(client, { from, to })
+    /* 
+    - f0 sends 8 to f4
+    - f4 sends 2 to f0
+
+         f0, f4
+    f0 [ 0,  8 ] 
+    f4,[ 2,  0 ]
+    */
+    const entityIndex = {}
+    const names = []
+    // first pass to find the unique entities / size of the matrix
+    for (const m of messages) {
+      let fromIndex = entityIndex[m.from]
+      if (!fromIndex) {
+        fromIndex = names.length
+        names[fromIndex] = m.from
+        entityIndex[m.from] = fromIndex
+      }
```

Comment:
I get uncomfortable managing array indexes so personally I'd use `push`. Also non-zero indexed arrays are unusual in JS so subtracting `-1` here (and using a null/undefined check) so `names` is zero indexed.

```suggestion
      let fromIndex = entityIndex[m.from]
      if (fromIndex == null) {
        fromIndex = names.push(m.from) - 1
        entityIndex[m.from] = fromIndex
      }
```

Reply:
yep, `names` was a last minute addition to make it simpler to track the names in the correct order, so was a little scrappy. Will change.